### ISSUE TITLE
FIFOQueue as a part of custom layer

### ISSUE THREAD

**Issue by markub3327** [keras-team/keras#16015]
Hello,

   I need a queue as a part of my custom input layer in Transformer for Short-term memory purposes. I need to store the coming features "in first in first out" manner and remove the oldies features when it's full and clean all content if it's needed. I see [here](https://www.tensorflow.org/api_docs/python/tf/queue/FIFOQueue) something like that for working with Tensor, but is it a good choice for me and contains all that I need? Is the FIFOQueue faster than using Python's deque or Numpy based queue directly in a custom layer?

Compare Python's deque collection with TF's FIFOQueue:
| Python | TensorFlow |
|--------|-------------|
| deque(maxlen=max_size) | tf.queue.FIFOQueue(capacity, dtypes, shapes=None, names=None, shared_name=None)
| append(x)<br># if it's full discard from the left end | enqueue(vals) |
| clear() | ??? |
| len(queue) | size() |
| pop() | dequeue() |
| queue[i] | ??? |

Or alternatively do it with Numpy like:
```python
queue = np.zeros((maxlen, features)) # my queue
x = np.random.normal(size=features) # new features

# append()
queue = np.concatenate([queue[:-1], x[np.newaxis, :]])

# predict
y = model(queue[np.newaxis, :, :])
```

With Python's timeit I get a benchmark on 100000 cycles with maxlen=1000 and features=6:
| Method | Time |
|--------|-------------|
| Numpy | 0.1765849579999994  s |
| deque | 12.360834958000002  s |

Thanks, have a nice day.

**Comment by LukeWood** [comment]
It may be better to write a low level custom training loop than a custom layer for this.

**Comment by LukeWood** [comment]
You may also want to use the `tf.TensorArray` API to implement this.  It has better support for random access/writing.

**Comment by markub3327** [comment]
@LukeWood 

The custom training loop cannot be used during deploying the model to the embedded device for measuring time-series sensor data that comes in one-by-one order. During training is an issue only with the number of batches. The example above works with only one batch on the layer output.

This layer transforms data during inference also training !!!

Thanks for the help, I'll try `tf.TensorArray`.

**Comment by markub3327** [comment]
@LukeWood 

 Please how can I replace effectively `self.queue.assign(tf.concat([self.queue[timesteps:, :], inputs], axis=0))` with methods of `tf.TensorArray`?
